### ISSUE TITLE
[Video Disruptors Grant]: Kinora— Enhanced Livepeer Video Player + SDK for AI Quests w/ Privacy Preserving Metrics

### ISSUE THREAD

**Issue by emmajane1313** [livepeer/grants#170]
### Please describe your project. Start with the need or problem you are trying to solve with this project. Describe why your solution is going to adequately solve this problem.

**Problem**
Web2 video players and platforms have for many years now offered useful engagement metrics, seamless social features, and elaborate interface customization options for developers. Yet, inherently, they lack the transparency, respect for user privacy, any semblance of creator equity, and the kind of creative monetization and engagement incentives only recently made possible by decentralization.

Creators, viewers, developers, and even brands remain dependent on channels that can cut you off from secure communication, your data, your ability to make informed choices in the market, your daily viewing habits, and your connections to other people through your social graph. It’s probably nothing.

**Solution**
(what) Our solution is to make it easier for developers to build a more regenerative ecosystem of web3 apps and services for video creation, distribution, viewership, and incentive models alike – using generative AI credits unsealed through creator and viewer quests.

Alongside a proof of concept, this proposal details two core technical components, an sdk and an extended player, with three main modules for privacy preserving metrics, channel engagement quests, and decentralized social media integration.

Critically, it will open access to AI text, image, audio, and video generation tooling, through GPU credits unlocked within participatory quests.

User activity produces user approved, token gated, and conditionally signed data to fine-tune generation for personalized content and merchandise.

When AI-content sells, it funds more GPU access for creators. This gives them better tools to make higher production quality content, without big upfront costs. It creates a loop where creators and the ecosystem feed each other. Sales drive incentives through quests, which in turn generates privacy preserving data about engagement, giving creators the means and opportunity to keep benefiting.

This cycle drives decentralized media forward.

(why) Eliminating the gap between web2 and web3 video creator, viewer, and developer experiences will speed up the timeline to when we can all leave our dependencies on archaic, centralized, and anti-user platforms behind.

### Link to demo website (if applicable)

https://chromadin.xyz

### Please describe in more detail why this proposal is valuable for the Livepeer ecosystem

Many creators still aren't ready to join web3 video or social media networks. Migrating ecosystems is a heavy lift for developers, too. It’s a big commitment.

Livepeer already solves the base video layer for what creators, viewers, and devs need to take the possibility of that switch away from web2 platform lock-in seriously. But there is more to build. Another layer is needed.

While Livepeer reduces transcoding costs and expands p2p markets for video delivery, this SDK and extended player component will help drive down costs, increase access, and make development and use easier for generative AI models in social video networks.

Today hyper-concentrations of power and resources in a handful of LLM operators block the way. VRAM, context length, model training, open dataset, and machine uptime friction dwarf even traditional video transcoding and delivery costs.

The risk of not building this layer is that Microsoft, Alphabet, X, and a few rival conglomerates will win. They will use their early lead to completely and permanently own the entire stack, from social commerce to LLM generation of everything we can make or imagine, code and content included.

The challenge to getting it right is staying targeted, flexible, and simple. Uncapped numbers of independent teams shipping in public, and quick. As long as we get that much right, the practical effects of open source and decentralization, backed by p2p market activity rooted in user autonomy, can compete head to head with old conglomerate platforms to overcome any early advantage.

Let markets decide who wins when open infrastructure is built, ready to scale.

### Please describe in details what your final deliverable for this project will be.

At the end of this project, we will deliver the SDK, a live extended player component, and a proof-of-concept (PoC) demo on both mobile and desktop dApps, all of which will be fully tested and documented.

**Kinora SDK:**

The SDK will bring together LitAuth, Programmable Key Pairs, and a custom on-chain Activity Assessment Module to record engagement metrics through the Livepeer player component, while keeping user privacy intact.

Users will authenticate via a connected wallet or web2 LitAuth Discord/Google auth. Their interactions within the Livepeer Player component will be logged by the Activity Assessment Module, which includes: 
- Average View Duration (AVD)
- Click-Through Rate (CTR)
- Social shares via Lens protocol
- Return Visitor Rate
- Bounce Rate

These metrics will be stored on-chain with IPFS. When Programmable Key Pairs are minted and granted, they'll automate the signing and dispatch of encrypted interaction batches to the Kinora Contract Suite, preserving valuable user data securely and reliably while maintaining full privacy in an efficient way. 

The Kinora SDK will also expose dedicated functions for connecting to the Kinora Contract Suite, allowing developers to easily connect metrics recorded by the Activity Assessment Module to on-chain quests, that when completed by users, unlock GPU credits and token-gated personalized content and merchandise. The quest logic will also support:
Ready deployment of custom quest logic, including participant approval criteria, requisite sub-milestones, and unsealed reward dispatch.

- User-friendly steps to help approved participants activate and begin a particular quest.
- Sub-milestones for each quest. 
     - As users fulfill the event conditions, completion is automatically validated. The user's unique Programmable Key Pair 
     signs the completion transaction, which is programmatically bundled with the pre-approved award tokens (NFTs or 
     ERC20s) at quest instantiation.
- On-chain metadata and other token-gated content.
    - Access to this content is granted only after users complete the necessary sub-milestones and hold the specified 
    decryption tokens.
- The ability for users to retrieve their quest history, along with current status and progress of any given quests.
- A view and claim option for unlocked compute credits, distributed by dispatchers assigned within the Contract Factory Suite.

The SDK will be built to support both node.js and standard browser implementations.

**Extended Functionality of the Player Component**
The Livepeer Player component will be enhanced by additional style props, callback props, Lens Protocol props, and instance methods. These improvements enable deeper media integration, more efficient state management, customized visual layout options, and a direct interface with corresponding Lens Protocol data.

Callback Props include: onReady, onPlay, onStart, onProgress, onDuration, onPause, onBuffer, onSeek, onEnded, and onError.

Style Props include: Width, Height, Style, and Controls.

Lens Protocol Props include: Publication Id, Total Amount of Upvotes, Total Amount of Downvotes, Total Amount of Collects, Total Amount of Mirrors, and Total Amount of Comments.

Instance Methods include: seekTo(), getSecondsLoaded(), getCurrentTime(), and getDuration().

**Proof-of-Concept (PoC):** 
The PoC will demonstrate a full-scale implementation of the Kinora SDK and the expanded Player component.

These components will be incorporated into Chromadin, a 24/7 LoFi channel, video streaming social platform, and NFT market built on the Lens Protocol, which has recently been deployed within the DIGITALAX ecosystem. Currently, Chromadin supports social media interactions, 24/7 video playlists, token-gated content, and a storefront where viewers can collect apparel and print NFTs from select creators.

With the SDK operational within the dApp, users get an incentivized and interactive engagement model. Authenticated users can receive personalized video recommendation feeds, complete quests, and earn awards in the form of NFTs, apparel, and generative model credits, all unlocked by their active participation and engagement.

In this PoC, all video streaming and playlists on Chromadin will utilize the newly extended Player component for both mobile and desktop interfaces.

### Please break up your development work into a clear set of milestones

**First Milestone ($5,000)**
-   Start development of the Kinora SDK.
-   Ensure compatibility for Node.js and web browsers.
-   Integrate PKP Mint/Grant functionality.
-   Add user wallet and Lit web2 authentication mechanism.
-   Build out Activity Assessment Module covering metrics: CTR, AVD, Bounce Rate, Lens Protocol Social Shares, and Return Visitor Rate.
-  Develop and unit test Kinora Smart Contract Factory Suite.
-  Set up IPFS storage and batch transaction system for on-chain signing with the Kinora Smart Contract Suite.
-  Run detailed functionality tests to confirm the code's reliability. 
**Final delivery:** Production ready and fully tested Kinora SDK, alongside the Quest Contract Factory. 
**Duration:** 3 weeks Team: Emma-Jane MacKinnon-Lee (Lead developer & validation tester)
**Dates:** August 15th 2023 - September 8th 2023

**Second Milestone ($2,000)**
-   Develop the extended Player Component, which includes callback, style props, and instance methods.
-   Add Lens Protocol publication Id statistics data.
-   Ensure all methods and props, including Instance Methods, Callback Props, and Style Props, are functional.
-   Conduct functionality tests to validate the code. 
**Final delivery:** Production ready and fully tested extended Player component.
**Duration:** 1 week Team: Emma-Jane MacKinnon-Lee (Lead developer & validation tester)
**Dates:** September 10th 2023 - September 15th 2023

**Third Milestone ($6,000)**
-   Integrate the Kinora SDK within Chromadin PoC and support its mobile application.
- Integrate the extended Player component for enhanced video playlists and streaming on Chromadin.
-   Connect Activity Assessment Module analytics to the Kinora SDK, facilitating user quests, apparel and NFT awards, and personal video suggestions.
-   Perform thorough functionality tests to verify the code's integrity. 
**Final Delivery:** Integrated SDK and extended Player component in the Chromadin dApp.
**Duration:** 4 weeks Team: Emma-Jane MacKinnon-Lee (Lead developer & validation tester)
**Dates:** September 17th 2023 - October 17th 2023

**Fourth Milestone ($2,000)**
-   Prepare detailed technical documentation and user flow diagrams.
-   Launch the SDK, the extended Player component, and Chromadin PoC.
**Final Delivery:** Detailed gitbook documentation.
**Duration:** 1 week Team: Emma-Jane MacKinnon-Lee (Technical writer & QA specialist)
**Dates:** October 19th 2023 - October 26th 2023

### Sum up the total requested budget across all milestones, and include that figure here. Also, please include a budget breakdown to specify how you are planning to spend these funds.

**Total Budget Requested:** USD $15,000

**Budget Breakdown:**
-   Development & Testing
-   Deployment Resources
-   Quality Assurance and Testing
-   Community bug bounty program (~$500) to motivate developer testing of the SDK and find potential issues.
-   Community Building
-   Content Development
-   User Support: Maintain support channels for the developer SDK and user engagement with the PoC. This will include current documentation, FAQ, a dedicated Discord channel, and clear video tutorials.
-   Developer and Creator Outreach.

### Specify your team's long-term plans to maintain this software and upgrade it over time

We are committed to building this SDK, the extended player, and the generative model quest layer because, like so many other creators and developers held back by web2 dependencies, we need it ourselves.

Development and maintenance will continue as it generates surplus time and resources, within our ongoing work.

And, critically, the incentives model reinforces growth of an active developer community building new applications on an irrevocably open source codebase, without worry of being priced out of future API access.

### Please describe (in words) your team's relevant experience, and why you think you are the right team to build this project. You can cite your team's prior experience in similar domains, doing similar dev work, individual team members' backgrounds, etc.

**Relevant Experience**
- Full stack TS + react + solidity + python + rust engineering
- Built on Ethereum for 5+ years
- Space engineering school dropout
- Co-led a crypto native hedge fund based out of Australia and the DIFC, Dubai, working on portfolio management, rebalancing, leading DeFi execution strategies and trading, particularly in options contracts across Ethereum and BTC for algorithmic variance swap strategy deployment
- Founder of DIGITALAX protocol + F3M, a web3 fashion synth line
- Ethonline 2022 finalist winner
- Developed a comprehensive Typescript SDK to automate the mint/grant/burn of Lit PKPs and conditional circuit logic, corresponding to on-chain events, webhooks, and timed intervals
- Built an all in one creative editor + decentralized social media app, Coin Op and Chromadin, with Stable Diffusion + Lens. 

**Team Members:**
Emma-Jane MacKinnon-Lee
https://github.com/digitalax
https://github.com/emmajane1313
https://digitalax.xyz/
https://chromadin.xyz/
https://f3manifesto.xyz/
https://themanufactory.xyz/

### Who is your target user group? How do you plan on getting your users to use this?

Applications built on the SDK and integrating the extended player are meant for creators, viewers, developers, and brands that remain locked in to web2 despite the litany of harmful anti-user behavior centralized social platforms continue to surprise us with. Users recognize how bad the status quo is, and why they should finally leave, but are still held back from making the move. 

The key to breaking through that blockade, in our opinion, is to make forming new social connections and viewing habits much easier, while reducing costs to creators, and clearly communicating reliable ways to collect autonomous incentives.

### How did you learn about the Livepeer Grants Program?

Since late 2022 we have been in full build mode. Every project we have built during this time has focused intensively on Lens Protocol integration, growing from the conviction that web2 social platforms are in free fall, but still drive commerce. That independence from market volatility can only come from building the new means of production, fulfillment, social interaction, and direct sales between creators, collectors, and emerging web3 communities. During this build cycle, it has been clear that video is far and away the single most important generator of market, network, and community activity. It has also been painfully clear how much friction there still is in implementing ready set video components.

A couple of months ago, while trying to switch Chromadin's video player to the Livepeer Player component, I hit a roadblock due to some limitations in Livepeer's existing functionality. This was mostly because of custom styling and triggered callback functions. I tried to solve these issues through a support ticket on the Discord, (which seems to have been removed since it was over 30 days old?). But, it branched into a fruitful conversation with Shih-Yu about how these not yet built features would be great if they were developed through a Livepeer grant so that others who need the same features can also benefit.  
 
Soon after, once work on another PKP related project was completed, the decision was made that this grant is the best way to build a better developer experience for ourselves and every other developer who sees the need for it a bit like we do.

### Was this project started at a hackathon or another web3 event? Which one?

The need for this project emerged during work on related apps chipping away at parallel aspects of why creators, collectors, and devs are still locked in to anti-user web2 platforms instead of migrating to autonomy enhancing networks and tools at a rate comparable to ChatGPT, Stable Diffusion, Threads, or any of the fastest growing apps from the web2 era.

### Please include any additional information that you think would be useful in helping us to evaluate your proposal.

There is a common thread that connects the SDK and extended player as key social video infrastructure to reinforce our work in:

- the DIGITALAX ecosystem (to unbundle fashion and micro-manufacturing through web3 creative commerce)
- Chromadin (a social video, NFT, print, and apparel distribution channel)
- The Dial (an all-in-one creative editor w/ full Lens and Stable Diffusion integration)
- The Manufactory (leading the rollout of autonomous print, apparel, and micro-manufacturing fulfillment networks)
- Coin-Op (a near instant AI streetwear and streetart print design to fulfillment app)
- the Lit Listener (a circuit design SDK and no-code interface for encrypted, conditional, and on-chain events)
- Legend (a new take on web3 social grants for public goods).

Each part of this work has a why in common: remove barriers to market for platform independent creators by combining attention mechanisms, user owned privacy, abstracted engagement metrics, social communication, generative models, and autonomous commerce.

**Comment by hansy** [comment]
Thanks for applying @emmajane1313! We'll get back to you with a decision within 2-4 weeks!

**Comment by hansy** [comment]
Hey @emmajane1313, we'd love to fund this project! Do you mind DMing me on Discord (@hansy_) so we can fine tune the milestones together? 

Also, I pinged some members of the Livepeer team responsible for the Livepeer Player regarding your proposed updates. Here's the message in the Livepeer Discord: https://discord.com/channels/423160867534929930/680142091174543362/1141710833894838302. 

I'm also keen to get their thoughts around your proposal around on-chain analytics, considering the Player also collects metrics.

**Comment by emmajane1313** [comment]
Awesome! will ping u on discord, looking forward to diving in

**Comment by hansy** [comment]
Thanks for the chat @emmajane1313! Based on our conversation we'll amend the milestones as follows:

**Core Milestones:**
Milestone 1: $4000
Milestone 2: $2000
Milestone 3: $4000
Milestone 4: $2000

**Success metric:** $3000
- 100 unique quest creations with SDK
- 1000 total quest completions

I'll queue up the 20% upfront payment; upon completion, I'll reply back with the tx receipt along with the dates of completion for all the milestones above.

**Comment by hansy** [comment]
Here's the tx receipt for the upfront payment: https://arbiscan.io/tx/0x229c1083642951d8390e1cb7f5b4cea1093e65e2b8bf35db9f5fe5763af1dec7.

With that, the grant is officially kicked off! Here are the milestone due dates:

Milestone 1: September 18, 2023
Milestone 2: September 25, 2023
Milestone 3: October 23, 2023
Milestone 4: October 30, 2023
Success metric: October 30, 2023

Let me know if you have any questions!

**Comment by emmajane1313** [comment]
Great! Getting started ✈️

**Comment by hansy** [comment]
Here's the tx receipt to Milestone 2: https://arbiscan.io/tx/0xec690f24c779e44ee5b38d1fe83fd92f994095acb80d64cdddf1a7de77b2d3f7! Let me know if you have any questions!

**Comment by emmajane1313** [comment]
Great! starting with milestone 2

**Comment by emmajane1313** [comment]
Replying to unmark the issue as stale. Have been communicating through discord with Hansy, Milestone 3 (The demo app) is still in active development. Designs have been shared and excited for a final release soon.
